PXB-2.4-1961 Prepare of full backup fails for encrypted tables with Gener…

rahulmalik87 opened the pull request:
…ic error

https://jira.percona.com/browse/PXB-1961

Problem:
PXB is not handling redo type generated by PS for encryption keyring

Analysis:
PS supports addition encryption types like keyring which is not handled
by PXB. This bugs only adresses encryption='n'

Fix:
Process KEY_MAGIC_PS generated by PS. also error out if redo log on page
0 found with CRYPT_SCHEME_1 that is keyring encryption is used

rahulmalik87 commented:
jenkin results https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-param-medium/176/

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
 		the data pages. */
 		if (page_no == 0 && !is_system_tablespace(space_id)) {
-			if (fil_write_encryption_parse(ptr, end_ptr,
+			byte* ptr_copy = ptr;
+			//increase by offset
+                        ptr_copy += 2;
+                        ulint len = mach_read_from_2(ptr_copy);
+                        ptr_copy += 2;
```
use comment
// skip offset

Also, alignment seems wrong from the comment

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+                        ptr_copy += 2;
+                        ulint len = mach_read_from_2(ptr_copy);
+                        ptr_copy += 2;
+                        if (end_ptr < ptr_copy + len)
+                                return NULL;
```
if (end_ptr < ptr_copy + len) {
                                return(NULL);
                   }

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+                        if (end_ptr < ptr_copy + len)
+                                return NULL;
+
+			/* to only support encryption ='n' with PS */
```
Add a bit more detailed comment
/* Tables created with ENCRYPTION='N' have crypt_data (UNENCRYPTED SCHEM) in Page 0 */

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+			/* to only support encryption ='n' with PS */
+			if (memcmp(ptr_copy, ENCRYPTION_KEY_MAGIC_PS_V1,
+                                   ENCRYPTION_MAGIC_SIZE) == 0) {
+				// increase by  magic size, iv_len, space_id, offset,
```
say // skip magic_size, iv_len, space_id, offset to read the "type" in crypt_data

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+			if (memcmp(ptr_copy, ENCRYPTION_KEY_MAGIC_PS_V1,
+                                   ENCRYPTION_MAGIC_SIZE) == 0) {
+				// increase by  magic size, iv_len, space_id, offset,
+				ptr_copy += 3 + 2  + 4 + 2 ;
```
better to use constants for these.
iv_len is 2 bytes?

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+				// increase by  magic size, iv_len, space_id, offset,
+				ptr_copy += 3 + 2  + 4 + 2 ;
+				uint type = mach_read_from_1(ptr_copy);
+				if(type != CRYPT_SCHEME_UNENCRYPTED) {
```
if (

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+				uint type = mach_read_from_1(ptr_copy);
+				if(type != CRYPT_SCHEME_UNENCRYPTED) {
+					ib::error() << "Can't take backup of "
+						"tablespace encrypted with "
+						"KEYRING encrypion";
```
better to add  "<<" at start of every line (some compilers don't like this concatenation)

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1852,28 @@ recv_parse_or_apply_log_rec_body(
+			byte* ptr_copy = ptr;
+			//increase by offset
+                        ptr_copy += 2;
+                        ulint len = mach_read_from_2(ptr_copy);
```
can you debug assert ut_ad(len == KEYRING_ENCRYPTION_SIZE)

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1869,37 @@ recv_parse_or_apply_log_rec_body(
 		the data pages. */
 		if (page_no == 0 && !is_system_tablespace(space_id)) {
-			if (fil_write_encryption_parse(ptr, end_ptr,
+			byte* ptr_copy = ptr;
+			//increase by offset
+                        ptr_copy += 2;
+                        ulint len = mach_read_from_2(ptr_copy);
```
This alignment still looks wrong to me. Please check your tabs vs spaces

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1625,6 +1625,23 @@ recv_scan_log_seg_for_backup(
 }
 #endif /* UNIV_HOTBACKUP */
 
+#ifdef XTRABACKUP
```
do we need the Ifdefs?

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1869,37 @@ recv_parse_or_apply_log_rec_body(
+                                return NULL;
+			}
+
+			/* Tables created with ENCRYPTION='N' have crypt_data
```
With Percona Server, tables created ..

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1869,37 @@ recv_parse_or_apply_log_rec_body(
+			}
+
+			/* Tables created with ENCRYPTION='N' have crypt_data
+			(UNENCRYPTED SCHEM) in Page 0 */
```
SCHEME

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1869,37 @@ recv_parse_or_apply_log_rec_body(
+				ut_ad(len == encrypt_info_size);
+				ptr_copy += 2;
+
+				//skip space_id, offset to read the type in crypt data
```
// skip

satya-bodapati commented on storage/innobase/xtrabackup/test/t/keyring_pxb_1961.sh:
```diff
@@ -0,0 +1,68 @@
+# PXB-1961: xtrabackup fails with encryption='n' with Percona server
+############################################################################
+
+function backup_and_restore() {
+  record_db_state test
+  rm -rf $topdir/backup
+  mkdir -p $topdir/backup
+  # Backup
+  run_cmd xtrabackup --backup --target-dir=$topdir/backup --xtrabackup-plugin-dir=${plugin_dir}
+
+  # Prepare
+  run_cmd $XB_BIN $XB_ARGS --xtrabackup-plugin-dir=${plugin_dir} --prepare \
+  --target-dir=$topdir/backup ${keyring_args}
+
+  # Restore
+  stop_server
+  rm -rf $mysql_datadir
+  run_cmd xtrabackup --copy-back --target-dir=$topdir/backup
+  start_server
+  # Verify backup
+  verify_db_state test
+}
```
this seems very useful. Can you make it a bit generic by passing database as input parameter.

and move it to common.sh

satya-bodapati replied:
@rahulmalik87 can you please address this comment as well?

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1869,37 @@ recv_parse_or_apply_log_rec_body(
+
+				uint iv_len = mach_read_from_2(ptr_copy);
+				uint encrypt_info_size = fil_get_encrypt_info_size(iv_len);
+				ut_ad(len == encrypt_info_size);
```
this is fine,
I also want to compare with hard-coded numbers. (I belive only two lengths are possible)

rahulmalik87 commented:
jenkin results. https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-param-medium/176/

satya-bodapati commented on storage/innobase/xtrabackup/test/t/keyring_pxb_1961.sh:
```diff
@@ -0,0 +1,68 @@
+  rm -rf $topdir/backup
+  mkdir -p $topdir/backup
+  # Backup
+  run_cmd xtrabackup --backup --target-dir=$topdir/backup --xtrabackup-plugin-dir=${plugin_dir}
```
--xtrabackup-plugin-dir is necessary? :thinking:

satya-bodapati replied:
may be lets keep it. Useful or keyring plugin stuff I guess.

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1868,39 @@ recv_parse_or_apply_log_rec_body(
+			}
+
+                        /* With Percona Server, Tables created with
+			ENCRYPTION='N' have crypt_data (UNENCRYPTED SCHEMA)
```
it is just SCHEME

satya-bodapati reviewed (changes requested):
patch looks good. can you please submit to Jenkins and update the link here?

rahulmalik87 commented:
> patch looks good. can you please submit to Jenkins and update the link here?

https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-param-medium/181/testReport/

satya-bodapati reviewed (changes requested):
1. can you rename test to avoid numbers? (Please remember this. Applicable for all PRs)
may be keyring_encryption.sh ?
2. I see that the test is skipped on debug builds also.
https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-param-medium/181/BUILD_TYPE=debug,Host=min-centos-8-x64,xtrabackuptarget=innodb57/consoleFull

`keyring_pxb_1961                         w4	[skipped]   Require debug server version`

rahulmalik87 commented:
> 1. can you rename test to avoid numbers? (Please remember this. Applicable for all PRs)
>    may be keyring_encryption.sh ?
> 2. I see that the test is skipped on debug builds also.
>    https://pxb.cd.percona.com/job/percona-xtrabackup-2.4-param-medium/181/BUILD_TYPE=debug,Host=min-centos-8-x64,xtrabackuptarget=innodb57/consoleFull
> 
> `keyring_pxb_1961 w4 [skipped] Require debug server version`

renamed the testcase, waiting for fix for PXB-2512. paste the manual result

{smm-12-d40aec x7 o7 xtrabackup-test}$./run.sh -t t/keyring_encryption.sh
Using /home/rahul.malik/MySQL/build/x7/xtrabackup-test/var as test root
Detecting server version...
Running against MySQL 5.7.34-debug-log (InnoDB 5.7.34)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
keyring_encryption                       w1	[skipped]   Requires XtraDB
==============================================================================
Spent 26 of 33 seconds executing testcases

SUMMARY: 1 run, 0 successful, 1 skipped, 0 failed

Skipped tests:  keyring_encryption

See results/ for detailed output
{smm-12-d40aec x7 o7 xtrabackup-test}$n xp7
{smm-12-d40aec x7 p7 xtrabackup-test}$lnb
{smm-12-d40aec x7 p7 xtrabackup-test}$./run.sh -t t/keyring_encryption.sh
Using /home/rahul.malik/MySQL/build/x7/xtrabackup-test/var as test root
Detecting server version...
Running against Percona Server 5.7.33-36-debug-log (XtraDB 5.7.33-36)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
keyring_encryption                       w1	[passed]    28
==============================================================================
Spent 28 of 36 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

See results/ for detailed output

satya-bodapati reviewed (approved):
LGTM!

Commit:
PXB-1961 Prepare of full backup fails for encrypted tables with Generic error

https://jira.percona.com/browse/PXB-1961

Problem:
PXB is not handling redo type generated by PS for encryption keyring

Analysis:
PS supports addition encryption types like keyring which is not handled
by PXB. This bugs only adresses encryption='n'

Fix:
Process KEY_MAGIC_PS generated by PS. also error out if redo log on page
0 found with CRYPT_SCHEME_1 that is keyring encryption is used

rahulmalik87 commented:
./run.sh -t t/keyring_encryption.sh
Using /home/rahul.malik/MySQL/build/x7/xtrabackup-test/var as test root
Detecting server version...
Running against Percona Server 5.7.33-36-debug-log (XtraDB 5.7.33-36)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
keyring_encryption                       w1	[passed]    29
==============================================================================
Spent 29 of 37 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

See results/ for detailed output
{smm-12-d40aec x7 o7 xtrabackup-test}$

satya-bodapati commented on storage/innobase/log/log0recv.cc:
```diff
@@ -1852,7 +1868,38 @@ recv_parse_or_apply_log_rec_body(
+				return NULL;
+			}
+
+			/* With Percona Server, Tables created with
+                                        ENCRYPTION='N' have crypt_data
+                             (CRYPT_SCHEME_UNENCRYPTED) in Page 0 */
```
Please align the comment :)

rahulmalik87 commented:
./run.sh -t t/keyring_encryption.sh
Using /home/rahul.malik/MySQL/build/x7/xtrabackup-test/var as test root
Detecting server version...
Running against Percona Server 5.7.33-36-debug-log (XtraDB 5.7.33-36)
Using 'xtrabackup' as xtrabackup binary
Using 8 parallel workers

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
keyring_encryption                       w1	[passed]    35
==============================================================================
Spent 35 of 44 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

See results/ for detailed output
{smm-12-d40aec x7 o7 xtrabackup-test}$
